core/modules/replica/RebalanceJob.cc

```diff
@@ -649,39 +583,109 @@ void RebalanceJob::onJobFinish(MoveReplicaJob::Ptr const& job) {
 
+    for (auto&& ptr: _jobs) {
+        if (numAtDest.end() == numAtDest.find(ptr->destinationWorker())) {
+            numAtDest[ptr->destinationWorker()] = 0;
+        }
+        if (numAtSrc.end() == numAtSrc.find(ptr->sourceWorker())) {
+            numAtSrc[ptr->sourceWorker()] = 0;
+        }
+    }
+    for (auto&& ptr: _activeJobs) {
+        auto itrAtDest = numAtDest.find(ptr->destinationWorker());
+        if (numAtDest.end() == itrAtDest) {
+            numAtDest[ptr->destinationWorker()] = 1;
         } else {
-            finish(lock, ExtendedState::FAILED);
+            itrAtDest->second++;
+        }
+        auto itrAtSrc = numAtSrc.find(ptr->sourceWorker());
+        if (numAtSrc.end() == itrAtSrc) {
+            numAtSrc[ptr->sourceWorker()] = 1;
+        } else {
+            itrAtSrc->second++;
         }
     }
-}
-
-void RebalanceJob::release(unsigned int chunk) {
+    
+    // Try to fulfil the request (to submit the given number of jobs)
+    // by evaluating best candidates using an algorithm explained
+    // within the loop below.
+    
+    size_t numJobsLaunched = 0;
+    for (size_t i = 0; i < numJobs; ++i) {
+
+        // THE LOAD BALANCING ALGORITHM:
+        //
+        //   The algorithms evaluates candidates (pairs of (dstWorker,srcWorker))
+        //   to find the one which allows more even spread of load among the destination
+        //   and source workers. For each pair of the workers the algorithm computets
+        //   a 'load' which is just a sum of the on-going activities at both ends of
+        //   the proposed transfer:
+        //
+        //     load := numAtDest[destWorker] * numAtSrc[srcWorker]
+        //
+        //   A part which has the lowest number will be selected.
+
+        size_t minLoad = ULLONG_MAX;
+        MoveReplicaJob::Ptr job;
+
+        for (auto&& ptr: _jobs) {            
+            size_t const load = numAtDest[ptr->destinationWorker()] +
+                                numAtSrc [ptr->sourceWorker()];
+            if (load <= minLoad) {
+                minLoad = load;
+                job = ptr;
+            }
+        }
+        if (nullptr != job) {
 
-    // THREAD-SAFETY NOTE: This method is thread-agnostic because it's trading
-    // a static context of the request with an external service which is guaranteed
-    // to be thread-safe.
+            // Update occupancy of the worker nodes at both ends
+            numAtDest[job->destinationWorker()]++;
+            numAtSrc [job->sourceWorker()]++;
 
-    LOGS(_log, LOG_LVL_DEBUG, context() << "release  chunk=" << chunk);
+            // Move the job into another queue
+            _activeJobs.push_back(job);
+            _jobs.remove(job);
 
-    Chunk chunkObj {databaseFamily(), chunk};
-    controller()->serviceProvider()->chunkLocker().release(chunkObj);
+            // Let it run
+            job->start();
+            numJobsLaunched++;
+        }
+    }
+    return numJobsLaunched;
```

Comment:
This `launchNextJobs()` looks very similar to `ReplicateJob::launchNextJobs()`, try to reduce code duplication.

Reply:
You're right. And not only this method. The whole machinery for submitting requests in a controlled way which was introduced with this PR is significantly duplicated in implementation of both classes (RebalanceJob and ReplicateJob). And I already thought about coming up with a template version of request submitting algorithm. Chances are that I may also need the for other kind of requests. However, this will require some design effort, and I'm not sure if I should do this now, or register it in my TODO list and do it later.